### PR TITLE
New resource azurerm_managed_disk_sas_token to manage disk exports

### PR DESCRIPTION
**Pull request by harshavmb**
New resource azurerm_disk_export to manage disk exports.
This PR is inspired from  [azurerm_storage_account_blob_container_sas](https://github.com/harshavmb/terraform-provider-azurerm/blob/a22742cf74baa2cd932e72caebbb3382f57870dd/internal/services/storage/storage_account_blob_container_sas_data_source.go) datasource

Few details about disk access grants here :: https://docs.microsoft.com/en-us/rest/api/compute/disks/grant-access
https://docs.microsoft.com/en-us/rest/api/compute/disks/revoke-access

With the help of this disk export sas token, we could copy data from disks onto storage blobs or elsewhere without the need of azcopy

Looking forward to hear from you.

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @harshavmb - it looks like there are some issues with the test:

```
[](https://ci-oss.hashicorp.engineering/viewLog.html?buildId=260323&buildTypeId=TerraformOpenSource_TerraformProviders_AzureRMPublic_AZURERM_SERVICE_PUBLIC_COMPUTE#)------- Stdout: -------
=== RUN   TestAccDataSourceDiskExportSas_basic
=== PAUSE TestAccDataSourceDiskExportSas_basic
=== CONT  TestAccDataSourceDiskExportSas_basic
    testcase.go:110: Step 1/1 error: Check failed: Check 1/4 error: data.azurerm_managed_disk_export.test: Attribute 'managed_disk_id' expected "/subscriptions/42cbb0b8a331-abaf-4e69-8d8f-14b86a40/resourceGroups/disksrg/providers/Microsoft.Compute/disks/disk1", got "/subscriptions/*******/resourceGroups/acctestRG-disk-220223000403491059/providers/Microsoft.Compute/disks/acctestsadsq8vx2"
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: deleting Managed Disk "acctestsadsq8vx2" (Resource Group "acctestRG-disk-220223000403491059"): compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="OperationNotAllowed" Message="There is an active shared access signature outstanding for disk acctestsadsq8vx2. Call EndGetAccess before attaching or deleting the disk. Learn more here: aka.ms/revokeaccessapi."
        
--- FAIL: TestAccDataSourceDiskExportSas_basic (45.57s)
FAIL
```

**Comment by harshavmb** [comment]
> Thanks @harshavmb - it looks like there are some issues with the test:
> 
> ```
> [](https://ci-oss.hashicorp.engineering/viewLog.html?buildId=260323&buildTypeId=TerraformOpenSource_TerraformProviders_AzureRMPublic_AZURERM_SERVICE_PUBLIC_COMPUTE#)------- Stdout: -------
> === RUN   TestAccDataSourceDiskExportSas_basic
> === PAUSE TestAccDataSourceDiskExportSas_basic
> === CONT  TestAccDataSourceDiskExportSas_basic
>     testcase.go:110: Step 1/1 error: Check failed: Check 1/4 error: data.azurerm_managed_disk_export.test: Attribute 'managed_disk_id' expected "/subscriptions/42cbb0b8a331-abaf-4e69-8d8f-14b86a40/resourceGroups/disksrg/providers/Microsoft.Compute/disks/disk1", got "/subscriptions/*******/resourceGroups/acctestRG-disk-220223000403491059/providers/Microsoft.Compute/disks/acctestsadsq8vx2"
>     testing_new.go:70: Error running post-test destroy, there may be dangling resources: exit status 1
>         
>         Error: deleting Managed Disk "acctestsadsq8vx2" (Resource Group "acctestRG-disk-220223000403491059"): compute.DisksClient#Delete: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="OperationNotAllowed" Message="There is an active shared access signature outstanding for disk acctestsadsq8vx2. Call EndGetAccess before attaching or deleting the disk. Learn more here: aka.ms/revokeaccessapi."
>         
> --- FAIL: TestAccDataSourceDiskExportSas_basic (45.57s)
> FAIL
> ```

Hi @katbyte ,

Thanks for the review. Indeed, it was something I missed. I added another data source to revoke the granted access on the disk. 
Tested locally, it worked. Please rereview this..

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @harshavmb 

Thanks for this PR

Taking a look through here since these are creating and subsequently revoking resources on the Azure side these want to be a single Resource rather than two Data Sources here - so can we update this to be a single Resource rather than two Data Sources?

Thanks!

**Comment by harshavmb** [comment]
> hi @harshavmb
> 
> Thanks for this PR
> 
> Taking a look through here since these are creating and subsequently revoking resources on the Azure side these want to be a single Resource rather than two Data Sources here - so can we update this to be a single Resource rather than two Data Sources?
> 
> Thanks!

Hi @tombuildsstuff 

I've added a new resource as you suggested. Changes LGTM, tested locally. 

Please re-review.

**Comment by harshavmb** [comment]
Hi @tombuildsstuff @katbyte @jackofallops ,

Could you please provider some update on this PR? Waiting in the review queue for quite sometime.

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @harshavmb

Thanks for this PR - apologies for the delayed re-review here.

I've taken a look through and left some more comments inline, in particular I can't see this defined as a "Disk Export" within Azure, since this is a SAS Token for a Managed Disk, I'm thinking that this resource would make more sense with that name  -WDYT?

Thanks!

File: internal/services/compute/disk_export_resource.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		return fmt.Errorf("parsing Managed Disk ID %q: %+v", parsedManagedDiskId.ID(), err)
+	}
+
+	diskName := parsedManagedDiskId.DiskName
+	resourceGroupName := parsedManagedDiskId.ResourceGroup
+
+	grantAccessData := compute.GrantAccessData{
+		Access:            compute.AccessLevel(access),
+		DurationInSeconds: &durationInSeconds,
+	}
```
[the remark this answers is not in the document]
Comment (harshavmb) [reply internal/services/compute/disk_export_resource.go:88]: Hi @tombuildsstuff ,

I think disk access is more related to private links. I did check the API again [here](https://docs.microsoft.com/en-us/rest/api/compute/disks/get). `diskAccessId` existed only for disks with private endpoints. 

Having said that, I found `diskState` parameter which will be `ActiveSAS` when the disk export is active. I'll check this parameter.

File: internal/services/compute/disk_export_resource.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		Delete: resourceDiskExportDelete,
+
+		Timeouts: &pluginsdk.ResourceTimeout{
+			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
+			Update: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*pluginsdk.Schema{
```
[the remark this answers is not in the document]
Comment (harshavmb) [reply internal/services/compute/disk_export_resource.go:35]: Hmm., I don't think tags would be of much user here. I couldn't find them in the API documentation [here](https://docs.microsoft.com/en-us/rest/api/compute/disks/grant-access). 
Please let me know if you need those tags.

File: internal/services/compute/disk_export_resource.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+			"managed_disk_id": {
+				Type:             pluginsdk.TypeString,
+				Required:         true,
+				DiffSuppressFunc: suppress.CaseDifference,
+				ValidateFunc:     azure.ValidateResourceID,
+			},
+
+			"duration_in_seconds": {
+				Type:     pluginsdk.TypeInt,
+				Required: true,
```
[the remark this answers is not in the document]
Comment (harshavmb) [reply internal/services/compute/disk_export_resource.go]: The azure limits are between 30 & 4294967295. Apparently, the higher limit 4294967295 breaks 32-bit compatibility tests as it breaches the upper limits of int. 

Please advice.

File: internal/services/compute/disk_sas_token_resource.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-11-01/compute"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/compute/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/compute/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
+)
+
+func resourceDiskExport() *pluginsdk.Resource {
```
Comment (katbyte) [inline internal/services/compute/disk_sas_token_resource.go:17]: Could we change function names to match the new resource name?

File: internal/services/compute/disk_sas_token_resource_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+resource "azurerm_managed_disk" "test" {
+  name                 = "acctestsads%s"
+  location             = azurerm_resource_group.test.location
+  resource_group_name  = azurerm_resource_group.test.name
+  storage_account_type = "Standard_LRS"
+  create_option        = "Empty"
+  disk_size_gb         = "1"
+}
+resource "azurerm_managed_disk_sas_token" "test" {
```
Comment (katbyte) [inline internal/services/compute/disk_sas_token_resource_test.go:67]:
```suggestion
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-revokedisk-%d"
  location = "%s"
}

resource "azurerm_managed_disk" "test" {
  name                 = "acctestsads%s"
  location             = azurerm_resource_group.test.location
  resource_group_name  = azurerm_resource_group.test.name
  storage_account_type = "Standard_LRS"
  create_option        = "Empty"
  disk_size_gb         = "1"
}

resource "azurerm_managed_disk_sas_token" "test" {
```

File: website/docs/r/disk_export.html.markdown [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+subcategory: "Compute"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_managed_disk_sas_token"
+description: |-
+  Manages a Disk Export.
+---
+
+# azurerm_managed_disk_sas_token
+
+Manages a Disk Export.
```
Comment (katbyte) [inline website/docs/r/disk_export.html.markdown:11]: and can we make the name changes throughout the docs?

File: website/docs/r/disk_export.html.markdown [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+* `sas_url` - The computed Shared Access Signature (SAS) of the Managed Disk.
+
+## Timeouts
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/docs/configuration/resources.html#timeouts) for certain actions:
+
+* `create` - (Defaults to 30 minutes) Used when creating the Disk.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Disk.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Disk.
```
Comment (katbyte) [inline website/docs/r/disk_export.html.markdown:68]: we should include how to import the resource here?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks, just some minor comments i've left inline and this should be good!

File: website/docs/r/disk_export.html.markdown (hunk 7 above)
Comment (harshavmb) [reply website/docs/r/disk_export.html.markdown:68]: The SAS tokens are ephemeral, I don't know how useful it would be to import the disk here.

**Comment by harshavmb** [comment]
> Thanks, just some minor comments i've left inline and this should be good!

Thank you, I've made changes as requested. Let me know if there is something I need to change

**Pull request merged**